File: local_preview_server.py
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import unquote
import mimetypes
import os


ROOT = Path(__file__).resolve().parent


class IMSIPreviewHandler(SimpleHTTPRequestHandler):
    def translate_path(self, path: str) -> str:
        path = unquote(path.split("?", 1)[0].split("#", 1)[0])
        if path == "/":
            rel = "index.html"
        elif path.startswith("/imsi/"):
            rel = path[len("/imsi/"):]
        else:
            rel = path.lstrip("/")

        target = (ROOT / rel).resolve()
        try:
            target.relative_to(ROOT)
        except ValueError:
            return str(ROOT)
        return str(target)

    def end_headers(self) -> None:
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def guess_type(self, path):
        mime, _ = mimetypes.guess_type(path)
        if mime:
            return mime
        return "application/octet-stream"


def main() -> None:
    port = int(os.environ.get("IMSI_PREVIEW_PORT", "4000"))
    server = ThreadingHTTPServer(("127.0.0.1", port), IMSIPreviewHandler)
    print(f"Serving IMSI preview at http://127.0.0.1:{port}/imsi/")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: local_preview_server.rb
require "webrick"
require "pathname"

ROOT = Pathname.new(__dir__).realpath
PORT = 4000

server = WEBrick::HTTPServer.new(
  Port: PORT,
  BindAddress: "127.0.0.1",
  AccessLog: [],
  Logger: WEBrick::Log.new($stderr, WEBrick::Log::WARN)
)

server.mount_proc "/" do |req, res|
  path = req.path
  relative =
    if path == "/"
      "index.html"
    elsif path.start_with?("/imsi/")
      path.delete_prefix("/imsi/")
    else
      path.sub(%r{\A/}, "")
    end

  target = ROOT.join(relative).cleanpath

  unless target.to_s.start_with?(ROOT.to_s) && target.exist?
    res.status = 404
    res["Content-Type"] = "text/plain; charset=utf-8"
    res.body = "Not Found"
    next
  end

  target = target.join("index.html") if target.directory?

  unless target.exist? && target.file?
    res.status = 404
    res["Content-Type"] = "text/plain; charset=utf-8"
    res.body = "Not Found"
    next
  end

  res.status = 200
  res["Cache-Control"] = "no-store"
  res["Content-Type"] = WEBrick::HTTPUtils.mime_type(target.extname, WEBrick::HTTPUtils::DefaultMimeTypes)
  res.body = target.binread
end

trap("INT") { server.shutdown }
trap("TERM") { server.shutdown }

puts "Serving IMSI preview at http://127.0.0.1:#{PORT}/imsi/"
server.start
